### PR TITLE
stage2: bringup plan9 target and linker

### PR DESCRIPTION
**Pull request by g-w1**
* [x] sucessful `pub fn main() void {}`
* [x] Convert to use a global offset table (or just some fancy relocations)
* [x] sucessful hello world (This may be harder as the pwrite syscall needs arguments on the stack, so I might have to implement more inline asm which may be out of scope). ~~Out of scope, MOV for inline assembly will take more work in stage2.~~
* [x] use correct code for 64 bit targets (not `@intCast`) to u32
* [x] remove some panics/TODOs
scrot

Makes strides towards #7153.

### CODE REVIEW

**Comment by kubkon** [comment]
How much is left (is it viable) for having plan9 autotested in stage2 tests?

**Comment by g-w1** [comment]
Very good question! Sourcehut has support for plan9 on ci https://sourcehut.org/blog/2020-05-11-sourcehut-plus-plan-9/, so we could probably use that, but maybe not until its self-hosting or we can generate the binaries on a different platform, then run them on the plan9 ci.
Russ Cox also made a thing, kind of like wine, that can run plan9 user programs on some unixes. https://swtch.com/9vx/ Maybe we could use this on the ci? (I haven't used it yet, I'll try it soon)

**Comment by kubkon** [comment]
> Very good question! Sourcehut has support for plan9 on ci https://sourcehut.org/blog/2020-05-11-sourcehut-plus-plan-9/, so we could probably use that, but maybe not until its self-hosting or we can generate the binaries on a different platform, then run them on the plan9 ci.
> Russ Cox also made a thing, kind of like wine, that can run plan9 user programs on some unixes. https://swtch.com/9vx/ Maybe we could use this on the ci? (I haven't used it yet, I'll try it soon)
> Edit: Someone on irc also said that you could just boot plan9 in qemu and run the program in it.

Hmm, I guess srht would be our best bet since rolling out wine-like emulation or QEMU might be a tough thing to do inside another VM (i.e., the CI). Perhaps it could be done though?

**Comment by g-w1** [comment]
Successful hello world!!!: 
image

**Comment by g-w1** [comment]
For reviewers: here is the link to the man page used for reference http://git.9front.org/plan9front/plan9front/819bec22ccf01927c8d31e9bbcddca91b5ed41ee/sys/man/6/a.out/f.html

File: lib/std/zig.zig [hunk 1]
```diff
@@ -178,8 +178,32 @@ pub fn binNameAlloc(allocator: *std.mem.Allocator, options: BinNameOptions) erro
         .spirv => return std.fmt.allocPrint(allocator, "{s}.spv", .{root_name}),
         .hex => return std.fmt.allocPrint(allocator, "{s}.ihex", .{root_name}),
         .raw => return std.fmt.allocPrint(allocator, "{s}.bin", .{root_name}),
+        .plan9 => return std.fmt.allocPrint(allocator, "{s}.{c}", .{ root_name, archToPlan9Char(target.cpu.arch) }),
     }
 }
+fn archToPlan9Char(arch: std.Target.Cpu.Arch) ?u8 {
```
Comment (andrewrk) [inline lib/std/zig.zig:184]:
```suggestion
fn archToPlan9Char(arch: std.Target.Cpu.Arch) u8 {
```

File: src/link.zig [hunk 2]
```diff
@@ -328,6 +335,7 @@ pub const File = struct {
             .elf => return @fieldParentPtr(Elf, "base", base).updateDeclLineNumber(module, decl),
             .macho => return @fieldParentPtr(MachO, "base", base).updateDeclLineNumber(module, decl),
             .c => return @fieldParentPtr(C, "base", base).updateDeclLineNumber(module, decl),
+            .plan9 => @panic("PLAN 9 DEBUG INFO"),
```
Comment (andrewrk) [inline src/link.zig:338]:
```suggestion
            .plan9 => @panic("TODO: implement updateDeclLineNumber for plan9"),
```

File: src/link/Plan9.zig [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+const codegen = @import("../codegen.zig");
+const trace = @import("../tracy.zig").trace;
+const mem = std.mem;
+const File = link.File;
+const Allocator = std.mem.Allocator;
+
+const log = std.log.scoped(.link);
+const assert = std.debug.assert;
+
+// TODO use incremental compilation
```
Comment (andrewrk) [inline src/link/Plan9.zig:17]:
```suggestion
//! This implementation does all the linking work in flush(). A future improvement
//! would be to add incremental linking in a similar way as ELF does.
```

File: src/link/plan9/a.out.zig [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
```
Comment (andrewrk) [inline src/link/plan9/a.out.zig:20]: this zig code is not copyright-owned by plan 9 foundation; you can and should omit this copyright notice.

**Review by andrewrk — CHANGES_REQUESTED** [review]
I have only cosmetic changes requested on this. Impressive work!

**Pull request merged**